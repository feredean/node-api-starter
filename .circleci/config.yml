# Javascript Node CircleCI 2.0 configuration file
#
# Check https://circleci.com/docs/2.0/language-javascript/ for more details
#
version: 2.1

jobs:
  # build the project and prepare it for deployment
  build:
    docker:
      - image: circleci/node:12.20.1
    working_directory: ~/api
    steps:
      - checkout
      - restore_cache:
          keys:
            - v1-dependencies-{{ checksum "package.json" }}
            # fallback to using the latest cache if no exact match is found
            - v1-dependencies-
      - run: npm install
      - run: npm rebuild
      # cache node modules
      - save_cache:
          key: v1-dependencies-{{ checksum "package.json" }}
          paths:
            - node_modules
      - run: npm run build
      # cache distribution and files needed for deployment
      - save_cache:
          key: v1-build-{{ .Environment.CIRCLE_SHA1 }}
          paths:
            - dist
            - package-lock.json
            - package.json
            - Dockerfile

  # run tests using the cached dependencies
  test:
    docker:
      - image: circleci/node:12.20.1
      - image: circleci/mongo:4.4
    working_directory: ~/api
    steps:
      - checkout
      - restore_cache:
          keys:
            - v1-dependencies-{{ checksum "package.json" }}
            - v1-dependencies-
      - run: npm test

  # build the image and push it to docker hub
  dockerize:
    machine: true
    working_directory: ~/api
    environment:
      IMAGE_NAME: feredean/node-api-starter
    steps:
      - restore_cache:
          keys:
            - v1-build-{{ .Environment.CIRCLE_SHA1 }}
      - run:
          name: Push image to docker registry
          command: |
            docker build -t $IMAGE_NAME:$CIRCLE_SHA1 -t $IMAGE_NAME:latest .
            echo "$DOCKERHUB_PASS" | docker login -u "$DOCKERHUB_USERNAME" --password-stdin
            docker push $IMAGE_NAME:latest
            docker push $IMAGE_NAME:$CIRCLE_SHA1

  # pull the image from docker hub and deploy it to the k8s cluster
  deploy:
    docker:
      - image: feredean/circleci-kops:0.1.0
    environment:
      IMAGE_NAME: feredean/node-api-starter
      KOPS_STATE_STORE: s3://k8s-explabs-io-state-store
    steps:
      - run:
          name: Deploy to k8s cluster
          command: |
            # Ensure AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are set in the project's env vars
            kops export kubecfg --name k8s.explabs.io
            kubectl set image deploy/node-api-starter node-api-starter=$IMAGE_NAME:$CIRCLE_SHA1

workflows:
  version: 2
  build_test_deploy:
    jobs:
      - build
      - test:
          requires:
            - build
      - dockerize:
          requires:
            - build
            - test
          filters:
            # only deploy when a commit hits the master branch
            branches:
              only: master
      - deploy:
          requires:
            - dockerize
          filters:
            branches:
              only: master
