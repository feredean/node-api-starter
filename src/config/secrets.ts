import dotenv from "dotenv";
import logger from "util/logger";
import { TEST, PRODUCTION, NODE_ENV } from "config/settings";
import fs from "fs";

if (!fs.existsSync(".env")) {
    logger.info("No .env file found, looking for variables in environment.");
}

if (NODE_ENV !== TEST) dotenv.config();

const secrets = [
    "SESSION_SECRET",

    "MONGO_DATABASE",
    "MONGO_USERNAME",
    "MONGO_PASSWORD",
    "MONGO_HOST",
    "MONGO_PORT",

    "FACEBOOK_ID",
    "FACEBOOK_SECRET",

    "SENDGRID_USER",
    "SENDGRID_PASSWORD",

    "AWS_ACCESS_KEY_ID",
    "AWS_ACCESS_KEY_SECRET",

    "CORS"
];

for (const secret of secrets) {
    if (!process.env[secret]) {
        logger.error(`Env variable ${secret} is missing.`);
        process.exit(1);
    }
}

let mongoURI;

if (NODE_ENV === PRODUCTION) {
    mongoURI = `mongodb+srv://${process.env.MONGO_USERNAME}:${process.env.MONGO_PASSWORD}@${process.env.MONGO_HOST}/${process.env.MONGO_DATABASE}?retryWrites=true&w=majority`;
} else {
    mongoURI = `mongodb://${process.env.MONGO_HOST}:${process.env.MONGO_PORT}/${process.env.MONGO_DATABASE}`;
}

export const SESSION_SECRET = process.env["SESSION_SECRET"];
export const MONGO_URI = mongoURI;
export const FACEBOOK_ID = process.env["FACEBOOK_ID"];
export const FACEBOOK_SECRET = process.env["FACEBOOK_SECRET"];
export const SENDGRID_USER = process.env["SENDGRID_USER"];
export const SENDGRID_PASSWORD = process.env["SENDGRID_PASSWORD"];
export const AWS_ACCESS_KEY_ID = process.env["AWS_ACCESS_KEY_ID"];
export const AWS_ACCESS_KEY_SECRET = process.env["AWS_ACCESS_KEY_SECRET"];
export const CORS = process.env["CORS"];