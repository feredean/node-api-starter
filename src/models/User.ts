import crypto from "crypto";
import bcrypt from "bcrypt";
import mongoose from "mongoose";
import UUID from "uuid/v4";
import { USER_ROLES } from "../config/settings";

export interface Profile {
    name?: string;
    gender?: string;
    location?: string;
    website?: string;
    picture?: string;
}
export interface AuthToken {
    accessToken: string;
    kind: string;
}
export interface UserAPIFormat {
    id: string;
    email: string;
    role: string;
    profile?: Profile;
    avatar: string;
}
export type UserDocument = mongoose.Document & {
    id: string;
    email: string;
    password: string;
    passwordResetToken: string;
    passwordResetExpires: Date;
    role: string;
    profile: Profile;

    facebook: string;
    tokens: AuthToken[];

    authenticate: (candidatePassword: string) => Promise<boolean>;
    gravatar: (size: number) => string;
    format: () => UserAPIFormat;
};

const userSchema = new mongoose.Schema(
    {
        id: { type: String, default: UUID, unique: true },
        email: { type: String, unique: true },
        password: String,
        passwordResetToken: String,
        passwordResetExpires: Date,
        role: { type: String, default: "user", enum: USER_ROLES },

        facebook: String,
        tokens: Array,

        profile: {
            name: String,
            gender: String,
            location: String,
            website: String,
            picture: String
        }
    },
    { timestamps: true }
);

userSchema.pre("save", async function(next: Function): Promise<void> {
    const user = this as UserDocument;
    if (!user.isModified("password")) return next();
    try {
        const salt = await bcrypt.genSalt(10);
        user.password = await bcrypt.hash(user.password, salt);
        next();
    } catch (error) {
        next(error);
    }
});

userSchema.methods = {
    authenticate: async function(candidatePassword: string): Promise<boolean> {
        return bcrypt.compare(candidatePassword, this.password);
    },
    gravatar: function(size: number = 200): string {
        if (!this.email) {
            return `https://gravatar.com/avatar/?s=${size}&d=retro`;
        }
        const md5 = crypto
            .createHash("md5")
            .update(this.email)
            .digest("hex");
        return `https://gravatar.com/avatar/${md5}?s=${size}&d=retro`;
    },
    format: function(): UserAPIFormat {
        const result = {
            id: this.id,
            email: this.email,
            role: this.role,
            avatar: this.gravatar(),
            profile: {
                name: this.profile.name,
                gender: this.profile.gender,
                location: this.profile.location,
                website: this.profile.website,
                picture: this.profile.picture
            }
        };

        return result;
    }
};

export const User = mongoose.model<UserDocument>("User", userSchema);
