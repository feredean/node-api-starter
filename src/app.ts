import express from "express";
import mongoose from "mongoose";
import setupExpress from "config/express";
import setupRoutesV1 from "config/routes/v1";
import { MONGO_URI } from "config/secrets";
import * as mw from "middleware";
import logger from "util/logger";

mongoose.set("useCreateIndex", true);
mongoose.set("useUnifiedTopology", true);

mongoose.connect(MONGO_URI, { useNewUrlParser: true })
    .catch((err): void => {
        logger.error("MongoDB connection error. Please make sure MongoDB is running. " + err);
        process.exit(1);
    });

const app = express();

setupExpress(app);
setupRoutesV1(app);

app.use(mw.handleMissing);
app.use(mw.handleErrors);

export default app;
