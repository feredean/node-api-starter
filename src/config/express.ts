import cors from "cors";
import morgan from "morgan";
import compression from "compression";
import bodyParser from "body-parser";
import { Express } from "express";

import { APP_PORT } from "./settings";
import { CORS_REGEX } from "./secrets";
import { setupPassport } from "./passport";
import logger from "../util/logger";

export const setupExpress = (app: Express): void => {
    app.set("port", APP_PORT);

    const corsOptions = {
        origin: (origin: string, callback: Function): void => {
            if (!origin) return callback();
            const match = origin.match(new RegExp(CORS_REGEX)) ? true : false;
            callback(null, match);
        }
    };

    setupPassport(app);

    app.use(cors(corsOptions));
    app.use(bodyParser.json());
    app.use(bodyParser.urlencoded({ extended: true }));
    app.use(compression());
    app.use(
        morgan(
            "[:method] :url :status :res[content-length] - :response-time ms",
            {
                stream: {
                    write: (text: string): void => {
                        logger.info(text.substring(0, text.lastIndexOf("\n")));
                    }
                }
            }
        )
    );
};
