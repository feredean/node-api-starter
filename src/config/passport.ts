import passport from "passport";
import passportLocal from "passport-local";
import passportFacebook, { Profile } from "passport-facebook";
import { Express, Request, Response, NextFunction } from "express";
import { User, UserDocument } from "../models/User";
import { FACEBOOK_ID, FACEBOOK_SECRET } from "./secrets";


const LocalStrategy = passportLocal.Strategy;
const FacebookStrategy = passportFacebook.Strategy;

/**
 * Sign in using Email and Password.
 */
passport.use(new LocalStrategy({
    usernameField: "email",
    passwordField: "password"
}, async (email, password, done): Promise<void> => {
    try {
        const user = await User.findOne({ email: email.toLowerCase() });
        if (!user) return done(null, false, {message: "Email not registered"});
        if (!await user.authenticate(password)) return done(null, false, { message: "Invalid credentials" });
        done(null, user);
    } catch (error) {
        return done(error);
    }
}));

/**
 * OAuth Strategy Overview
 *
 * - User is already logged in.
 *   - Check if there is an existing account with a provider id.
 *     - If there is, return an error message. (Account merging not supported)
 *     - Else link new OAuth account with currently logged-in user.
 * - User is not logged in.
 *   - Check if it's a returning user.
 *     - If returning user, sign in and we are done.
 *     - Else check if there is an existing account with user's email.
 *       - If there is, return an error message.
 *       - Else create a new account.
 */


/**
 * Sign in with Facebook.
 */
// passport.use(new FacebookStrategy({
//     clientID: FACEBOOK_ID,
//     clientSecret: FACEBOOK_SECRET,
//     callbackURL: "/auth/facebook/callback",
//     profileFields: ["name", "email", "link", "locale", "timezone"],
//     passReqToCallback: true
// }, async (req: Request, 
//     accessToken: string, 
//     refreshToken: string, 
//     profile: Profile,
//     done: Function
// ): Promise<void> => {
//     try {
//         if (req.user) {
//             let user = await User.findOne({ facebook: profile.id });
//             if (user) return done(undefined, false, { message: "There is already a Facebook account that belongs to you. Sign in with that account or delete it, then link it with your current account." });
//             user = await User.findById(req.user.id);
//             user.facebook = profile.id;
//             user.tokens.push({ kind: "facebook", accessToken });
//             user.profile.name = user.profile.name || `${profile.name.givenName} ${profile.name.familyName}`;
//             user.profile.gender = user.profile.gender || profile._json.gender;
//             user.profile.picture = user.profile.picture || `https://graph.facebook.com/${profile.id}/picture?type=large`;
//             await user.save();
//             done(undefined, user);
//         } else {
//             let user = await User.findOne({ facebook: profile.id });
//             if (user) return done(undefined, user);
//             user = await User.findOne({ email: profile._json.email });
//             if (user) return done(undefined, false, { message: "There is already an account using this email address. Sign in to that account and link it with Facebook manually from Account Settings." });
//             const newUser: UserDocument = new User();
//             newUser.email = profile._json.email;
//             newUser.facebook = profile.id;
//             newUser.tokens.push({ kind: "facebook", accessToken });
//             newUser.profile.name = `${profile.name.givenName} ${profile.name.familyName}`;
//             newUser.profile.gender = profile._json.gender;
//             newUser.profile.picture = `https://graph.facebook.com/${profile.id}/picture?type=large`;
//             newUser.profile.location = (profile._json.location) ? profile._json.location.name : "";
//             await newUser.save();
//             done(undefined, newUser);
//         }
//     } catch (error) {
//         done(error);
//     }
// }));

export const setupPassport = (app: Express): void => {
    app.use(passport.initialize());
};