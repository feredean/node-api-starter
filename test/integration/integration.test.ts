/* eslint-disable @typescript-eslint/no-explicit-any */
const mockedPutObject = jest.fn();
const mockedGetSignedUrl = jest.fn();
jest.mock("aws-sdk/clients/s3", (): any => {
    return class S3 {
        public putObject(params: any): any {
            mockedPutObject(params);
            return {
                promise: async (): Promise<void> => {}
            };
        }
        public getSignedUrl(operation: string, params: any): string {
            mockedGetSignedUrl(operation, params);
            return "https://dummy.url.com";
        }
    };
});

jest.mock("nodemailer");
import nodemailer from "nodemailer";
import bcrypt from "bcrypt";
import request from "supertest";
import jwt from "jsonwebtoken";

import {} from "types/passport-user";

import { RegisterUserOptions, registerValidUser, signToken, REGISTER_VALID } from "../helpers";
import { initMongo, disconnectMongo } from "../setup";
import { User } from "../../src/models/User";
import { SESSION_SECRET } from "../../src/config/secrets";

const GENERIC_UPLOAD_USER_ID = "GENERIC_UPLOAD_USER_ID";

interface JWTPayload {
    email: string;
    role: string;
    sub: string;
    exp: number;
}

import app from "app";
describe("API V1", (): void => {

    describe("/hello", (): void => {
        describe("GET /", (): void => {
            it("should return 200 OK", async (): Promise<void>  => {
                const { body } = await request(app).get("/v1/hello")
                    .expect(200);
                expect(body).toMatchSnapshot();
            });
        });
    });

    describe("/upload", (): void => {
        const adminOpts: RegisterUserOptions = {
            id: GENERIC_UPLOAD_USER_ID,
            role: "admin",
            randomize: true,
        };

        const userOpts: RegisterUserOptions = {
            id: GENERIC_UPLOAD_USER_ID,
            randomize: true,
        };

        describe("POST /", (): void => {
            beforeEach(async (): Promise<void> => {
                mockedPutObject.mockClear();
                mockedGetSignedUrl.mockClear();
                await initMongo();
            });

            afterAll(async(): Promise<void> => disconnectMongo());

            it("should return 201, the document added to storage and call s3.putObject, s3.getSignedUrl - one file", async (): Promise<void> => {
                const token = await registerValidUser(adminOpts);
                const { body } = await request(app)
                    .post("/v1/upload")
                    .set("authorization", `Bearer ${token}`)
                    .attach("file", "test/integration/files/sample-white.png")
                    .expect(201);

                expect(body).toMatchSnapshot();
                expect(mockedPutObject).toHaveBeenCalled();
                expect(mockedGetSignedUrl).toHaveBeenCalled();
            });

            it("should return 201, the documents added to storage and call s3.putObject, s3.getSignedUrl twice - two files", async (): Promise<void> => {
                const token = await registerValidUser(adminOpts);
                const { body } = await request(app)
                    .post("/v1/upload")
                    .set("authorization", `Bearer ${token}`)
                    .attach("file", "test/integration/files/sample-white.png")
                    .attach("file", "test/integration/files/sample-black.png")
                    .expect(201);
                
                expect(body).toMatchSnapshot();
                expect(mockedPutObject).toHaveBeenCalledTimes(2);
                expect(mockedGetSignedUrl).toHaveBeenCalledTimes(2);
            });

            it("should return 401 and not call s3.putObject, s3.getSignedUrl", async (): Promise<void> => {
                const token = await registerValidUser(userOpts);
                await request(app)
                    .post("/v1/upload")
                    .set("authorization", `Bearer ${token}`)
                    .attach("file", "test/integration/files/sample-white.png")
                    .attach("file", "test/integration/files/sample-black.png")
                    .expect(403);
                expect(mockedPutObject).toHaveBeenCalledTimes(0);
                expect(mockedGetSignedUrl).toHaveBeenCalledTimes(0);
            });
        });
    });

    describe("/users", (): void => {
        const adminOpts: RegisterUserOptions = { role: "admin" };

        describe("GET /", (): void => {
            beforeEach(async (): Promise<void> => {
                await initMongo();
                await registerValidUser(adminOpts);
            });
            afterAll(async (): Promise<void> => disconnectMongo());

            it("should return a list of users", async (): Promise<void> => {
                const user = await User.findOne({});
                const { body } = await request(app)
                    .get("/v1/users/")
                    .set("authorization", `Bearer ${signToken({
                        email: user.email,
                        id: user.id,
                        role: user.role,
                        exp: "1s"
                    })}`)
                    .expect(200);
                expect(body).toMatchSnapshot();
            });
        });
    });

    describe("/account", (): void => {
        const userOpts: RegisterUserOptions = {};
        
        describe("GET /jwt/refresh", (): void => {
            beforeEach(async (): Promise<void> => initMongo());
            afterAll(async (): Promise<void> => disconnectMongo());

            it("should return a fresher JWT", async (): Promise<void> => {
                const token = await registerValidUser(userOpts);
                const payload = await jwt.verify(token, SESSION_SECRET) as JWTPayload;
                const expiringToken = signToken({
                    id: payload.sub,
                    email: payload.email,
                    role: "user",
                    exp: "1s"
                });
                const refresh = await request(app)
                    .get("/v1/account/jwt/refresh")
                    .set("authorization", `Bearer ${expiringToken}`);

                const freshPayload = await jwt.verify(refresh.body.token, SESSION_SECRET) as JWTPayload;
                const oldPayload = await jwt.verify(expiringToken, SESSION_SECRET) as JWTPayload;
                expect(freshPayload.exp > oldPayload.exp).toBe(true);
            });

            it("should return 401 - token is valid but the user does not exist", async (): Promise<void> => {
                const token = await registerValidUser(userOpts);
                await User.deleteMany({});
                const refresh = await request(app)
                    .get("/v1/account/jwt/refresh")
                    .set("authorization", `Bearer ${token}`);
                expect(refresh.status).toBe(401);
            });
        });

        describe("POST /register", (): void => {

            const REGISTER_INVALID_EMAIL_PASSWORD = {
                "email": "a@a.a",
                "password": "pass"
            };

            const REGISTER_VALID_NAME_EMAIL_PASSWORD = {
                email: "valid@email.com",
                password: "valid_password",
                name: "valid name"
            };

            beforeEach(async (): Promise<void> => initMongo());
            afterAll(async (): Promise<void> => disconnectMongo());
        
            it("should return status 201, create a user and return a JWT - valid register", async (): Promise<void> => {
                const response = await request(app)
                    .post("/v1/account/register")
                    .send(REGISTER_VALID);

                expect(response.status).toBe(201);
                const payload = await jwt.verify(response.body.token, SESSION_SECRET);
                expect((payload as JWTPayload).email).toBe(REGISTER_VALID.email);
            });
        
            it("should return status 201, create a user with a name", async (): Promise<void> => {
                const response = await request(app)
                    .post("/v1/account/register")
                    .send(REGISTER_VALID_NAME_EMAIL_PASSWORD);

                expect(response.status).toBe(201);
                const user = await User.findOne({email: REGISTER_VALID_NAME_EMAIL_PASSWORD.email});
                expect(user.profile.name).toBe(REGISTER_VALID_NAME_EMAIL_PASSWORD.name);
            });
        
            it("should return status 422 - invalid email and password", async (): Promise<void> => {
                const { body } = await request(app)
                    .post("/v1/account/register")
                    .send(REGISTER_INVALID_EMAIL_PASSWORD)
                    .expect(422);
                expect(body).toMatchSnapshot();
            });
        
            it("should return status 422 - duplicate account", async (): Promise<void> => {
                await request(app).post("/v1/account/register").send(REGISTER_VALID);
                const { body } = await request(app)
                    .post("/v1/account/register")
                    .send(REGISTER_VALID)
                    .expect(422);
                expect(body).toMatchSnapshot();
            });
        
        });
     
        describe("POST /login", (): void => {  
            beforeEach(async (): Promise<void> => initMongo());
            afterAll(async (): Promise<void> => disconnectMongo());
            
            it("should return status 200 and the user's JWT - valid login", async (): Promise<void> => {
                await registerValidUser(userOpts);
                const response = await request(app)
                    .post("/v1/account/login")
                    .send(REGISTER_VALID);
                expect(response.status).toBe(200);
                const payload = await jwt.verify(response.body.token, SESSION_SECRET);
                expect((payload as JWTPayload).email).toBe(REGISTER_VALID.email);
            });

            it("should return status 403 - email not registered", async (): Promise<void> => {
                const { body } = await request(app)
                    .post("/v1/account/login")
                    .send(REGISTER_VALID)
                    .expect(403);
                expect(body).toMatchSnapshot();                    
            });

            it("should return status 403 - invalid credentials", async (): Promise<void> => {
                await registerValidUser(userOpts);
                const { body } = await request(app)
                    .post("/v1/account/login")
                    .send({
                        email: REGISTER_VALID.email,
                        password: "wrong_password"
                    })
                    .expect(403);
                expect(body).toMatchSnapshot();   
            });

            it("should return status 403 - no payload", async (): Promise<void> => {
                await registerValidUser(userOpts);
                const { body } = await request(app)
                    .post("/v1/account/login")
                    .send({})
                    .expect(403);
                expect(body).toMatchSnapshot();   
            });

        });

        describe("POST /forgot", (): void => {
            let sendMailMock: jest.Mock;
            beforeEach(async (): Promise<void> => {
                sendMailMock = jest.fn();
                (nodemailer.createTransport as jest.Mock).mockReturnValue({"sendMail": sendMailMock});

                await initMongo();
                await registerValidUser(userOpts);
            });
            afterAll(async (): Promise<void> => disconnectMongo());
            
            it("should return 201, set a password reset token and send an email", async (): Promise<void> => {
                await request(app)
                    .post("/v1/account/forgot")
                    .send({ email: REGISTER_VALID.email })
                    .expect(201);
                
                const user = await User.findOne({});
                expect(user.passwordResetToken).toBeDefined();
                expect(sendMailMock).toHaveBeenCalled();
            });

            it("should return 422, does not set a password reset token or send an email - no data", async (): Promise<void> => {
                const { body } = await request(app)
                    .post("/v1/account/forgot")
                    .send({})
                    .expect(422);
                expect(body).toMatchSnapshot();
                const user = await User.findOne({});
                expect(user.passwordResetToken).toBeUndefined();
                expect(sendMailMock).toBeCalledTimes(0);
            });

            it("should return 404, does not set a password reset token or send an email - email not registered", async (): Promise<void> => {
                const { body } = await request(app)
                    .post("/v1/account/forgot")
                    .send({ email: "not@registered.email" })
                    .expect(404);
                expect(body).toMatchSnapshot();
                const user = await User.findOne({});
                expect(user.passwordResetToken).toBeUndefined();
                expect(sendMailMock).toBeCalledTimes(0);
            });
        });

        describe("POST /reset/:token", (): void => {
            let sendMailMock: jest.Mock;
            const RESET_PASSWORD_VALID = {
                password: "different_valid_password",
                confirm: "different_valid_password"
            };
            const RESET_PASSWORD_INVALID = {
                password: "pass",
                confirm: "different_password"
            };
            beforeEach(async (): Promise<void> => {
                sendMailMock = jest.fn();
                (nodemailer.createTransport as jest.Mock).mockReturnValue({"sendMail": sendMailMock});
                await initMongo();
                await User.create({
                    email: "valid@email.com",
                    password: "$2b$10$dn9jixNaX2WCvnVWBfW4aucSPTS41hzE9.A3n7QLPL4bkHQ.6eCqK",
                    id: "b27e8455-eac8-47c9-babc-8a8a6be5e4ae",
                    passwordResetExpires: new Date(Date.now() + 60 * 60 * 1000),
                    passwordResetToken: "4ec7149e1c5d55721a9cb2b069c22ac0"
                });
            });
            afterAll(async (): Promise<void> => disconnectMongo());

            it("should return 201 and reset the password", async(): Promise<void> => {
                let user = await User.findOne({});
                await request(app)
                    .post(`/v1/account/reset/${user.passwordResetToken}`)
                    .send(RESET_PASSWORD_VALID)
                    .expect(201);
                user = await User.findOne({});
                expect(sendMailMock).toHaveBeenCalled();
                expect(await user.authenticate(RESET_PASSWORD_VALID.password)).toBe(true);
            });

            it("should return 422 - password mismatch, password too short and invalid token", async(): Promise<void> => {
                const { body } = await request(app)
                    .post("/v1/account/reset/invalid_token")
                    .send(RESET_PASSWORD_INVALID)
                    .expect(422);
                expect(body).toMatchSnapshot();
                expect(sendMailMock).toBeCalledTimes(0);
            });

            it("should return 422 - expired token", async(): Promise<void> => {
                const user = await User.findOne({});
                user.passwordResetExpires = new Date(Date.now() - 1000);
                await user.save();
                const { body } = await request(app)
                    .post(`/v1/account/reset/${user.passwordResetExpires}`)
                    .send(RESET_PASSWORD_VALID)
                    .expect(422);
                expect(body).toMatchSnapshot();
                expect(sendMailMock).toBeCalledTimes(0);
            });
        });

        describe("GET /profile", (): void => {
            beforeEach(async (): Promise<void> => await initMongo());
            afterAll(async (): Promise<void> => disconnectMongo());

            it("should return 200 and the user's profile information", async (): Promise<void> => {
                const token = await registerValidUser(userOpts);
                const { body } = await request(app)
                    .get("/v1/account/profile")
                    .set("authorization", `Bearer ${token}`)
                    .expect(200);
                expect(body).toMatchSnapshot();
            });

            it("should return 401 - invalid authorization token", async (): Promise<void> => {
                await registerValidUser(userOpts);
                await request(app)
                    .get("/v1/account/profile")
                    .expect(401);
            });
        });

        describe("POST /profile", (): void => {
            beforeEach(async (): Promise<void> => await initMongo());
            afterAll(async (): Promise<void> => disconnectMongo());
            const PROFILE_DATA = {
                name: "Valid User",
                gender: "User",
                location: "Userland",
                website: "valid.user.com"
            };

            it("should return 200 and change the user's profile information", async (): Promise<void> => {
                const token = await registerValidUser(userOpts);
                await request(app)
                    .post("/v1/account/profile")
                    .set("authorization", `Bearer ${token}`)
                    .send(PROFILE_DATA)
                    .expect(200);
                const user = await User.findOne({});
                expect(user.profile.name).toBe(PROFILE_DATA.name);
                expect(user.profile.gender).toBe(PROFILE_DATA.gender);
                expect(user.profile.location).toBe(PROFILE_DATA.location);
                expect(user.profile.website).toBe(PROFILE_DATA.website);
            });

            it("should return 200 and return the profile", async (): Promise<void> => {
                const token = await registerValidUser(userOpts);
                const { body } = await request(app)
                    .post("/v1/account/profile")
                    .set("authorization", `Bearer ${token}`)
                    .send(PROFILE_DATA)
                    .expect(200);
                expect(body).toMatchSnapshot();
            });

            it("should return 401 - invalid authorization token", async (): Promise<void> => {
                await registerValidUser(userOpts);
                await request(app)
                    .post("/v1/account/profile")
                    .send(PROFILE_DATA)
                    .expect(401);
                const user = await User.findOne({});
                expect(user.profile.name).toBeUndefined();
                expect(user.profile.gender).toBeUndefined();
                expect(user.profile.location).toBeUndefined();
                expect(user.profile.website).toBeUndefined();
            });
        });

        describe("POST /password", (): void => {
            beforeEach(async (): Promise<void> => initMongo());
            afterAll(async (): Promise<void> => disconnectMongo());

            const VALID_PASSWORD_PAYLOAD = {
                password: "newValidPassword",
                confirm: "newValidPassword"
            };

            const INVALID_PASSWORD_PAYLOAD = {
                password: "short",
                confirm: "wrong"
            };

            it("should return 200 and change the user's password", async (): Promise<void> => {
                const token = await registerValidUser(userOpts);
                await request(app)
                    .post("/v1/account/password")
                    .set("authorization", `Bearer ${token}`)
                    .send(VALID_PASSWORD_PAYLOAD)
                    .expect(200);
                const user = await User.findOne({});
                expect(await bcrypt.compare(VALID_PASSWORD_PAYLOAD.password, user.password)).toBe(true);
            });

            it("should return 401 - invalid authorization token", async (): Promise<void> => {
                await registerValidUser(userOpts);
                await request(app)
                    .post("/v1/account/password")
                    .set("authorization", "Bearer invalid_token")
                    .send(VALID_PASSWORD_PAYLOAD)
                    .expect(401);
                const user = await User.findOne({});
                expect(await bcrypt.compare(VALID_PASSWORD_PAYLOAD.password, user.password)).toBe(false);
            });

            it("should return 422 - invalid data", async (): Promise<void> => {
                const token = await registerValidUser(userOpts);
                await request(app)
                    .post("/v1/account/password")
                    .set("authorization", `Bearer ${token}`)
                    .send(INVALID_PASSWORD_PAYLOAD)
                    .expect(422);
                const user = await User.findOne({});
                expect(await bcrypt.compare(VALID_PASSWORD_PAYLOAD.password, user.password)).toBe(false);
            });
        });

        describe("POST /delete", (): void => {
            beforeEach(async (): Promise<void> => initMongo());
            afterAll(async (): Promise<void> => disconnectMongo());

            it("should return 200 and delete the user", async (): Promise<void> => {
                const token = await registerValidUser(userOpts);
                await request(app)
                    .post("/v1/account/delete")
                    .set("authorization", `Bearer ${token}`)
                    .expect(200);
                expect(await User.findOne()).toBeNull();
            });

            it("should return 401 - invalid authorization token", async (): Promise<void> => {
                await registerValidUser(userOpts);
                await request(app)
                    .post("/v1/account/delete")
                    .expect(401);
                expect(await User.findOne()).toBeDefined();
            });
        });
    });
});