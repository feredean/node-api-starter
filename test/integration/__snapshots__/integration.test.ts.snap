// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API V1 /account GET /profile should return 200 and the user's profile information 1`] = `
Object {
  "avatar": "https://gravatar.com/avatar/cb7529c9a7c3297760ec76e41bf77d0a?s=200&d=retro",
  "email": "valid@email.com",
  "id": "GENERIC_USER_ID",
  "profile": Object {},
  "role": "user",
}
`;

exports[`API V1 /account POST /forgot should return 404, does not set a password reset token or send an email - email not registered 1`] = `
Object {
  "errors": Array [
    Object {
      "msg": "Email not found",
    },
  ],
}
`;

exports[`API V1 /account POST /forgot should return 422, does not set a password reset token or send an email - no data 1`] = `
Object {
  "errors": Array [
    Object {
      "msg": "Invalid data",
    },
  ],
}
`;

exports[`API V1 /account POST /login should return status 403 - email not registered 1`] = `
Object {
  "errors": Array [
    Object {
      "msg": "Email not registered",
    },
  ],
}
`;

exports[`API V1 /account POST /login should return status 403 - invalid credentials 1`] = `
Object {
  "errors": Array [
    Object {
      "msg": "Invalid credentials",
    },
  ],
}
`;

exports[`API V1 /account POST /login should return status 403 - no payload 1`] = `
Object {
  "errors": Array [
    Object {
      "msg": "Invalid credentials",
    },
  ],
}
`;

exports[`API V1 /account POST /profile should return 200 and return the profile 1`] = `
Object {
  "avatar": "https://gravatar.com/avatar/cb7529c9a7c3297760ec76e41bf77d0a?s=200&d=retro",
  "email": "valid@email.com",
  "id": "GENERIC_USER_ID",
  "profile": Object {
    "gender": "User",
    "location": "Userland",
    "name": "Valid User",
    "website": "valid.user.com",
  },
  "role": "user",
}
`;

exports[`API V1 /account POST /register should return status 422 - duplicate account 1`] = `
Object {
  "errors": Array [
    Object {
      "msg": "Account already exists",
    },
  ],
}
`;

exports[`API V1 /account POST /register should return status 422 - invalid email and password 1`] = `
Object {
  "errors": Array [
    Object {
      "msg": "Please enter a valid email address",
    },
    Object {
      "msg": "Password must be at least 8 characters long",
    },
  ],
}
`;

exports[`API V1 /account POST /reset/:token should return 422 - expired token 1`] = `
Object {
  "errors": Array [
    Object {
      "msg": "Invalid token",
    },
  ],
}
`;

exports[`API V1 /account POST /reset/:token should return 422 - password mismatch, password too short and invalid token 1`] = `
Object {
  "errors": Array [
    Object {
      "msg": "Password must be at least 8 characters long",
    },
    Object {
      "msg": "Passwords do not match",
    },
    Object {
      "msg": "Invalid token",
    },
  ],
}
`;

exports[`API V1 /upload POST / should return 201, the document added to storage and call s3.putObject, s3.getSignedUrl - one file 1`] = `
Object {
  "data": Array [
    Object {
      "key": "GENERIC_UPLOAD_USER_ID/3eaf17869bb51bf27bd7c91bc9853973",
      "url": "https://dummy.url.com",
    },
  ],
}
`;

exports[`API V1 /upload POST / should return 201, the documents added to storage and call s3.putObject, s3.getSignedUrl twice - two files 1`] = `
Object {
  "data": Array [
    Object {
      "key": "GENERIC_UPLOAD_USER_ID/3eaf17869bb51bf27bd7c91bc9853973",
      "url": "https://dummy.url.com",
    },
    Object {
      "key": "GENERIC_UPLOAD_USER_ID/edc564026997e2cb819c01522780365a",
      "url": "https://dummy.url.com",
    },
  ],
}
`;

exports[`API V1 /users GET / should return a list of users 1`] = `
Object {
  "data": Array [
    Object {
      "avatar": "https://gravatar.com/avatar/cb7529c9a7c3297760ec76e41bf77d0a?s=200&d=retro",
      "email": "valid@email.com",
      "id": "GENERIC_USER_ID",
      "profile": Object {},
      "role": "admin",
    },
  ],
}
`;
